flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -252,23 +262,35 @@ protected StreamTask(
 	 * Runs the stream-tasks main processing loop.
 	 */
 	private void run() throws Exception {
-		final ActionContext actionContext = new ActionContext();
+		final Mailbox mailbox = taskMailboxExecutor.getMailbox();
+		final ActionContext actionContext = new ActionContext(mailbox);
+
+		mailbox.open();
+		mailboxLoopRunning = true;
+
 		while (true) {
+
 			if (mailbox.hasMail()) {
 				Optional<Runnable> maybeLetter;
 				while ((maybeLetter = mailbox.tryTakeMail()).isPresent()) {
-					Runnable letter = maybeLetter.get();
-					if (letter == POISON_LETTER) {
+					if (!mailboxLoopRunning) {
```

Comment:
**NB:** you are silently ignoring the letter here.

Should there be an additional check for `mailboxLoopRunning` before calling `performDefaultAction` too?

Reply:
I think that can be fixed by moving the before taking letters. The purpose of this bit more complex logic is that we assume change of `mailboxLoopRunning` is always connected to a letter and with this we do not have to check this field on each loop iteration, but only when we process mail. Therefore no check is required again for the default action.